### PR TITLE
Darktheme

### LINKED ISSUE #1190: Dark/Night theme
Is there a dark/night theme for prarielearn? If not, I'd like to start working on it!

### PR DESCRIPTION
**Pull request by matttang7**
Closes #1190 

- Created additional table themes that has userid and theme number to query and decide which theme to load

- Added styled toggle button

### CODE REVIEW

File: pages/userSettings/userSettings.ejs [hunk 1]
```diff
@@ -2,6 +2,8 @@
 <html>
   <head>
     <%- include('../partials/head'); %>
+    <link href="https://cdn.jsdelivr.net/gh/gitbrent/bootstrap4-toggle@3.6.1/css/bootstrap4-toggle.min.css" rel="stylesheet">
+    <script src="https://cdn.jsdelivr.net/gh/gitbrent/bootstrap4-toggle@3.6.1/js/bootstrap4-toggle.min.js"></script>
```
Comment (mwest1066) [inline pages/userSettings/userSettings.ejs]: We can't include anything from a CDN because PL needs to work without a network.

File: pages/userSettings/userSettings.ejs [hunk 2]
```diff
@@ -69,7 +71,25 @@
...
+          <fieldset>        
+              <%if(checked) {%>
+              <input type="checkbox" name="dark_mode" value="0" id="darkid" checked = true checked data-toggle="toggle" 
+              data-on="🌙" data-off="☀" data-onstyle="outline-primary" data-offstyle="outline-warning">dark_mode<br>  
+              <%} else { %>   
+                <input type="checkbox" name="dark_mode" value="0" id="darkid" data-toggle="toggle" 
+                data-on="🌙" data-off="☀" data-onstyle="dark" data-offstyle="primary">dark_mode<br> 
+              <%}  %> 
+              <!-- <input type="checkbox" name="other_thing1" value="2">other_thing1<br>     
+              <input type="checkbox" name="other_thing2" value="3">other_thing2<br>    -->
```
Comment (mwest1066) [inline pages/userSettings/userSettings.ejs]: Remove commented-out code.

File: package.json [hunk 3]
```diff
@@ -19,6 +19,7 @@
         "blocked-at": "^1.2.0",
         "body-parser": "^1.19.0",
         "bootstrap": "^4.4.1",
+        "bootstrap4-toggle": "^3.6.1",
```
Comment (mwest1066) [inline package.json]: I'd rather not include novel UI elements if there is a standard HTML element we can use instead. For example, I think we can just use a regular checkbox or radio in this case?

File: server.js [hunk 4]
```diff
@@ -863,8 +863,15 @@ module.exports.insertDevUser = function(callback) {
         var params = {user_id};
         sqldb.query(sql, params, function(err, _result) {
             if (ERR(err, callback)) return;
-            callback(null);
+            var sql
+            = 'INSERT INTO theme (id, user_id, themenum)'
+            + ' VALUES ($user_id, $user_id, 1)';
```
Comment (mwest1066) [inline server.js]: Please put all SQL code in the `.sql` files.

File: pages/userSettings/userSettings.js [hunk 5]
```diff
@@ -37,34 +37,69 @@ router.post('/', (req, res, next) => {
             if (ERR(err, next)) return;
             res.redirect(req.originalUrl);
         });
-    } else {
+    }
+    else if(req.body.__action === 'update_settings'){
```
Comment (mwest1066) [inline pages/userSettings/userSettings.js]: Please follow code style of other code (e.g., spaces around keywords and parens).

File: pages/userSettings/userSettings.js [hunk 6]
```diff
@@ -37,34 +37,69 @@ router.post('/', (req, res, next) => {
...
+        if('dark_mode' in req.body){
+            var params = {
+                user_id: res.locals.authn_user.user_id,
+                dark_mode: 1,
```
Comment (mwest1066) [inline pages/userSettings/userSettings.js]: Is this parameter a binary switch (`darkmode = 0/1`) or a theme number? Sometimes we seem to be doing one and sometimes the other.

**Comment by mwest1066** [comment]
The test suite is currently failing.

**Comment by mwest1066** [comment]
Rather than having an explicit toggle, I'd prefer to follow the user's OS-level preferences with the `prefers-color-scheme: dark` media query.

File: package-lock.json [hunk 7]
```diff
@@ -4920,8 +4925,8 @@
             "from": "git://github.com/mathjax/MathJax-src.git",
             "requires": {
                 "esm": "^3.2.25",
-                "mj-context-menu": "^0.2.0",
-                "speech-rule-engine": "^3.0.0-beta.6"
+                "mj-context-menu": "^0.2.2",
+                "speech-rule-engine": "^3.0.0-beta.8"
```
Comment (mwest1066) [inline package-lock.json]: Are these changes needed?

File: pages/partials/head.ejs [hunk 8]
```diff
@@ -3,6 +3,9 @@
 <meta name="viewport" content="width=device-width, initial-scale=1">
 <title>PrairieLearn</title>
 <link href="/node_modules/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet" />
+<% if(locals.darkmode == 1){%>
+<link href="/node_modules/bootstrap/dist/css/bootstrap1.min.css" rel="stylesheet" />
```
Comment (mwest1066) [inline pages/partials/head.ejs]: The `node_modules` directory is automatically created by `npm ci`, so we can't put things in there by hand. Extra themes should be installed by npm (preferably) or be put in `/public/stylesheets` (if necessary).

File: pages/partials/head.ejs [hunk 9]
```diff
@@ -3,6 +3,9 @@
...
+<% }%>
```
Comment (mwest1066) [inline pages/partials/head.ejs]: I'm worried about including random bootstrap themes. These have a bad history of becoming unsupported. Is there any more standard way to get darkmode with bootstrap?

**Comment by mwest1066** [comment]
What are the impacts of darkmode on questions? Do custom question elements (like `pl-file-editor` and `pl-drawing`) work correctly?

**Comment by mwest1066** [comment]
I'm closing this PR for now. If anyone wants to address the issues above then please open a new PR based on these commits.

**Comment by echuber2** [comment]
The file editor uses ACE so it could be toggled to use a dark theme itself, but courses may already be styling that. The other features like drawing stuff, MathJax rendering, etc. may be troublesome.

**Comment by harsh183** [comment]
Well it can be a start I suppose? Marked as experimental or Beta and slowly things get better. Ace has a really easy setting for setting up dark mode.

**Comment by echuber2** [comment]
Yes, there's nothing stopping the ACE editor (wrapped by `pl-file-editor` in PL) from using one of its own dark themes. However courses can already style that when they inject the editor, so they may be overriding it. I think it would be good to add a dropdown in the corner of the file editor for selecting a color scheme. Later, we could add an option for "default" if PL overrides that.

**Comment by harsh183** [comment]
Sure. Even if there are some light elements just having a dark background with light text makes a large difference. That plus ace dark support should work for like 95% of questions as is (not many people do fancy stuff really).
